bot/utils/redis_cache.py

```diff
@@ -0,0 +1,408 @@
+
+    @staticmethod
+    def _from_typestring(key_or_value: Union[bytes, str], prefixes: _PrefixTuple) -> RedisKeyOrValue:
+        """Deserialize a typestring into a valid Redis type."""
+        # Stuff that comes out of Redis will be bytestrings, so let's decode those.
+        if isinstance(key_or_value, bytes):
+            key_or_value = key_or_value.decode('utf-8')
+
+        # Now we convert our unicode string back into the type it originally was.
+        for prefix, _type in prefixes:
+            if key_or_value.startswith(prefix):
+                return _type(key_or_value[len(prefix):])
+        raise TypeError(f"RedisCache._from_typestring only supports the following: {prefixes}.")
+
+    # Add some nice partials to call our generic typestring converters.
+    # These are basically methods that will fill in some of the parameters for you, so that
+    # any call to _key_to_typestring will be like calling _to_typestring with the two parameters
+    # at `prefixes` and `types_string` pre-filled.
+    #
+    # See https://docs.python.org/3/library/functools.html#functools.partialmethod
+    _key_to_typestring = partialmethod(_to_typestring, prefixes=_KEY_PREFIXES)
+    _value_to_typestring = partialmethod(_to_typestring, prefixes=_VALUE_PREFIXES)
+    _key_from_typestring = partialmethod(_from_typestring, prefixes=_KEY_PREFIXES)
+    _value_from_typestring = partialmethod(_from_typestring, prefixes=_VALUE_PREFIXES)
+
+    def _dict_from_typestring(self, dictionary: Dict) -> Dict:
+        """Turns all contents of a dict into valid Redis types."""
+        return {self._key_from_typestring(key): self._value_from_typestring(value) for key, value in dictionary.items()}
+
+    def _dict_to_typestring(self, dictionary: Dict) -> Dict:
+        """Turns all contents of a dict into typestrings."""
+        return {self._key_to_typestring(key): self._value_to_typestring(value) for key, value in dictionary.items()}
+
+    async def _validate_cache(self) -> None:
+        """Validate that the RedisCache is ready to be used."""
+        if self._namespace is None:
+            error_message = (
+                "Critical error: RedisCache has no namespace. "
+                "This object must be initialized as a class attribute."
+            )
+            log.error(error_message)
+            raise NoNamespaceError(error_message)
+
+        if self.bot is None:
+            error_message = (
+                "Critical error: RedisCache has no `Bot` instance. "
+                "This happens when the class RedisCache was created in doesn't "
+                "have a Bot instance. Please make sure that you're instantiating "
+                "the RedisCache inside a class that has a Bot instance attribute."
+            )
+            log.error(error_message)
+            raise NoBotInstanceError(error_message)
+
+        await self.bot.redis_ready.wait()
```

Comment:
I think there's a bit of an issue with relying on `await self.bot.redis_ready.wait()` after the redis session is closed in `Bot.close()`. Specifically, when the bot calls `self.redis_ready.clear()`, it sets the condition to False, and with how `Condition.wait()` works, it could end up indefinitely suspending the coroutine while waiting for the condition to be set to True. This would be an issue if `_validate_cache()` was called during or after `Bot.close()`.

I'm not overly concerned about something intentionally interacting with the redis session after it's been closed, but it is certainly possible that a redis session method that uses `_validate_cache()` is called while `Bot.close()` is in progress or just finished (specifically after `self.redis_ready.clear()`), such as when restarting the bot.

To address this, I would recommend having a flag in the bot's attributes such as `_redis_closed` that starts as False, setting it to True just before the bot calls `await self.redis_session.wait_closed()` in `Bot.close()`, and adding a check at the start of `_validate_cache()` that raises an exception when `_redis_closed` is True.

(To make it additionally safe, you could check `_redis_closed` with a lock, but I'm not certain that would be necessary.)

Reply:
addressed in b7c30d4